### PR TITLE
HIP support

### PR DESCRIPTION
**Pull request by camierjs**
This PR adds an initial support for AMD GPUs by the use of the HIP runtime API and kernel language.
On the cluster, it should be testable with this environment (`bash`):
- `module load opt && module load rocm/2.7`
- `export PATH=/opt/rocm/hip/bin:$PATH`
- `export LD_LIBRARY_PATH=/opt/rocm/lib64:/opt/rocm/hsa/lib:$LD_LIBRARY_PATH`
- `salloc -I -N 1 -p pgpu -t 720`
- `make distclean && make hip -j`
- examples: `make ex1 && ./ex1 -pa -d hip`, `make ex6 && ./ex6 -pa -d hip`

### CODE REVIEW

**Comment by tzanio** [comment]
@camierjs, can you please update the CHANGELOG to say that we have added support for AMD GPUs via HIP?

File: config/config.hpp.in [hunk 1]
```diff
@@ -125,6 +125,10 @@
 // Requires a CUDA compiler (nvcc).
 // #define MFEM_USE_CUDA
 
+// Build the GPU/HIP-enabled version of the MFEM library.
```
Comment (tzanio) [inline config/config.hpp.in]: This could maybe say "AMD GPU/HIP-enabled"

File: config/defaults.mk [hunk 2]
```diff
@@ -46,6 +46,10 @@ CUDA_FLAGS = -x=cu --expt-extended-lambda -arch=$(CUDA_ARCH)
 CUDA_XCOMPILER = -Xcompiler=
 CUDA_XLINKER   = -Xlinker=
 
+# HIP configuration options
+HIP_CXX = hipcc
+HIP_FLAGS = --amdgpu-target=gfx900 # link-time option
```
Comment (tzanio) [inline config/defaults.mk]: So that needs to change for a different GPU model?

File: general/device.cpp [hunk 3]
```diff
@@ -150,6 +151,19 @@ static void CudaDeviceSetup(const int dev, int &ngpu)
...
 }
 
+static void HipDeviceSetup(const int dev, int &ngpu)
+{
+#ifdef MFEM_USE_HIP
+   int deviceId;
+   MFEM_GPU_CHECK(hipGetDevice(&deviceId));
+   hipDeviceProp_t props;
+   MFEM_GPU_CHECK(hipGetDeviceProperties(&props, deviceId));
+   printf("info: running on device #%d %s\n", deviceId, props.name);
```
Comment (tzanio) [inline general/device.cpp]: Use 'mfem::out' or 'mfem::err' instead of 'printf'.

File: config/config.hpp.in (hunk 1 above)
Comment (camierjs) [reply config/config.hpp.in]: done

File: general/device.cpp (hunk 3 above)
Comment (camierjs) [reply general/device.cpp]: done

File: config/defaults.mk (hunk 2 above)
Comment (camierjs) [reply config/defaults.mk]: Yes, and I added these comments:
```
# The 'amdgpu' option allows to specify the AMD GPU processor:
# gfx600:tahiti, gfx700:kaveri, gfx701:hawaii, gfx801:carrizo, gfx900, gfx1010
```

File: fem/bilininteg_diffusion.cpp [hunk 4]
```diff
@@ -1054,7 +1054,7 @@ static void PADiffusionApply(const int dim,
    }
 #endif // MFEM_USE_OCCA
 
-   if (Device::Allows(Backend::RAJA_CUDA))
+   if (Device::Allows(Backend::RAJA_CUDA) || (Device::Allows(Backend::HIP)))
```
Comment (tzanio) [inline fem/bilininteg_diffusion.cpp]: Check if this is necessary

File: general/device.cpp [hunk 5]
```diff
@@ -150,6 +151,20 @@ static void CudaDeviceSetup(const int dev, int &ngpu)
...
 }
 
+static void HipDeviceSetup(const int dev, int &ngpu)
+{
+#ifdef MFEM_USE_HIP
+   int deviceId;
+   MFEM_GPU_CHECK(hipGetDevice(&deviceId));
+   hipDeviceProp_t props;
+   MFEM_GPU_CHECK(hipGetDeviceProperties(&props, deviceId));
+   mfem::out << "info: running on device #"
```
Comment (tzanio) [inline general/device.cpp]: Can we remove this output?

Comment (camierjs) [reply general/device.cpp]: yes, sure

**Comment by tzanio** [comment]
Please also update the `INSTALL` file

File: general/device.cpp [hunk 6]
```diff
@@ -33,14 +33,15 @@ occa::device occaDevice;
...
    Backend::OCCA_OMP, Backend::RAJA_OMP, Backend::OMP,
    Backend::OCCA_CPU, Backend::RAJA_CPU, Backend::CPU
 };
 
 // Backend names listed by priority, high to low:
 static const char *backend_name[Backend::NUM_BACKENDS] =
 {
-   "occa-cuda", "raja-cuda", "cuda", "occa-omp", "raja-omp", "omp",
+   "occa-cuda", "raja-cuda", "cuda", "hip", "occa-omp", "raja-omp", "omp",
    "occa-cpu", "raja-cpu", "cpu"
```
Comment (tzanio) [inline general/device.cpp]: Update also the documentation of `Device:: Configure`?

Comment (camierjs) [reply general/device.cpp]: Done

**Comment by tzanio** [comment]
Thank you @camierjs, awesome work!

**Comment by camierjs** [comment]
Should be good to take a last look @dmed256, thanks for the feedback!

**Comment by tzanio** [comment]
@v-dobrev -- we will like to merge this because of changes in the `dmpi` PR.

It is a relatively self-contained change, but if you have any concerns about it, we can fix it in master :).

**Pull request merged**